File: research/chform_cpp/test_normest.cpp
// Validate C++ norm estimation (norm_est.hpp) vs brute force / dense.
// Build & run:
//   clang++ -std=c++20 -O3 -o /tmp/chf_ne research/chform_cpp/test_normest.cpp && /tmp/chf_ne
#include "chform.hpp"
#include "norm_est.hpp"

#include <cmath>
#include <cstdio>
#include <random>
#include <vector>

using namespace chf;

static CHForm random_state(int n, std::mt19937_64& rng, int depth) {
    CHForm c(n);
    for (int d = 0; d < depth; ++d) {
        for (int q = 0; q < n; ++q) c.clifford_1q((int)(rng() % 6), q);
        for (int q = 0; q + 1 < n; ++q) if (rng() & 1) c.cx(q, q + 1); else if (n >= 2) c.cz(q, (q + 1) % n);
    }
    double ang = (rng() % 100000) / 100000.0 * 6.2831853;
    double mag = 0.3 + (rng() % 1000) / 1000.0 * 1.7;
    c.scale(std::polar(mag, ang));
    return c;
}

static std::vector<cd> equatorial_dense(const imat& A) {
    int n = (int)A.size();
    std::vector<cd> v((size_t)1 << n);
    for (uint64_t x = 0; x < (1ULL << n); ++x) {
        int e = 0;
        for (int a = 0; a < n; ++a) if ((x >> a) & 1) { e += A[a][a]; for (int b = a + 1; b < n; ++b) if ((x >> b) & 1) e += 2 * A[a][b]; }
        v[x] = ipow(e);
    }
    double s = std::pow(SQRT1_2, n);
    for (auto& z : v) z *= s;
    return v;
}

int main() {
    std::mt19937_64 rng(2024);

    // 1) Lemma 4 exp_sum vs brute force
    double w1 = 0;
    for (int t = 0; t < 3000; ++t) {
        int m = (int)(rng() % 7);
        imat B(m, std::vector<int>(m, 0));
        for (int a = 0; a < m; ++a) { B[a][a] = (int)(rng() & 3); for (int b = a + 1; b < m; ++b) { int v = (int)(rng() & 1); B[a][b] = v; B[b][a] = v; } }
        w1 = std::max(w1, std::abs(exp_sum(B) - exp_sum_direct(B)));
    }
    printf("[%s] Lemma 4 exp_sum vs brute force on 3000 matrices (max abs err %.2e)\n",
           w1 < 1e-9 ? "OK" : "FAIL", w1);

    // 2) Lemma 3 inner_equatorial vs dense <phi|phi_A>
    double w2 = 0;
    for (int t = 0; t < 2000; ++t) {
        int n = 1 + (int)(rng() % 5);
        CHForm phi = random_state(n, rng, 1 + (int)(rng() % 3));
        imat A = random_equatorial(n, rng);
        cd got = inner_equatorial(phi, A);
        auto pa = equatorial_dense(A);
        cd ref(0, 0);
        for (uint64_t x = 0; x < (1ULL << n); ++x) ref += std::conj(phi.amplitude(x)) * pa[x];
        w2 = std::max(w2, std::abs(got - ref));
    }
    printf("[%s] Lemma 3 <phi|phi_A> vs dense on 2000 states (max abs err %.2e)\n",
           w2 < 1e-9 ? "OK" : "FAIL", w2);

    // 3) Lemma 2 estimate_norm2 -> exact ||sum term||^2
    double worst_rel = 0;
    for (int t = 0; t < 20; ++t) {
        int n = 2 + (int)(rng() % 3);
        int k = 2 + (int)(rng() % 4);
        std::vector<CHForm> terms;
        for (int j = 0; j < k; ++j) terms.push_back(random_state(n, rng, 1 + (int)(rng() % 2)));
        std::vector<cd> psi((size_t)1 << n, cd(0, 0));
        for (auto& tm : terms) { auto sv = tm.statevector(); for (size_t x = 0; x < sv.size(); ++x) psi[x] += sv[x]; }
        double exact = 0; for (auto& z : psi) exact += std::norm(z);
        double est = estimate_norm2(terms, n, 6000, rng);
        worst_rel = std::max(worst_rel, std::abs(est - exact) / exact);
    }
    printf("[%s] Lemma 2 norm estimate vs exact on 20 low-rank states (worst rel err %.3f, L=6000)\n",
           worst_rel < 0.2 ? "OK" : "FAIL", worst_rel);
    return 0;
}

File: research/chform_backend/test_normest.py
"""Validate the native inner product / norm estimation (norm_est.py) against
direct dense computation.

Run: python -m research.chform_backend.test_normest
"""

from __future__ import annotations

import numpy as np

from .chform import CHForm
from .engine import LowRankState
from . import norm_est as ne


def _rand_chform(n, depth, rng):
    """A random stabilizer state as a single CH-form (random Clifford on |0>)."""
    c = CHForm(n)
    names = ["H", "S", "S_DAG", "X", "Y", "Z"]
    for _ in range(depth):
        for q in range(n):
            c.clifford_1q(str(rng.choice(names)), q)
        for q in range(0, n - 1, 2):
            (c.cx if rng.random() < 0.5 else c.cz)(q, q + 1)
    c.scale(np.exp(1j * rng.uniform(0, 2 * np.pi)) * rng.uniform(0.3, 2.0))
    return c


def test_exp_sum(trials=300):
    """Lemma 4 O(n^3) exponential sum vs brute force."""
    rng = np.random.default_rng(0)
    worst = 0.0
    for _ in range(trials):
        m = int(rng.integers(0, 7))
        B = np.zeros((m, m), dtype=np.int64)
        for a in range(m):
            B[a, a] = rng.integers(4)
            for b in range(a + 1, m):
                B[a, b] = B[b, a] = rng.integers(2)
        worst = max(worst, abs(ne.exp_sum(B) - ne.exp_sum_direct(B)))
    assert worst < 1e-9, f"exp_sum mismatch {worst}"
    print(f"[OK] Lemma 4 exponential sum vs brute force on {trials} matrices "
          f"(max abs err {worst:.2e})")


def test_inner_equatorial(trials=300):
    """Lemma 3 <phi|phi_A> vs materialised inner product."""
    rng = np.random.default_rng(1)
    worst = 0.0
    for _ in range(trials):
        n = int(rng.integers(1, 6))
        phi = _rand_chform(n, int(rng.integers(1, 4)), rng)
        A = ne.random_equatorial(n, rng)
        got = ne.inner_equatorial(phi, A)
        ref = np.vdot(phi.statevector(), ne.equatorial_state(A))  # <phi|phi_A>
        worst = max(worst, abs(got - ref))
    assert worst < 1e-9, f"inner_equatorial mismatch {worst}"
    print(f"[OK] Lemma 3 <phi|phi_A> vs dense on {trials} states "
          f"(max abs err {worst:.2e})")


def test_norm_estimate(trials=12):
    """Lemma 2 norm estimator -> true ||psi||^2 (Clifford+T low-rank states)."""
    rng = np.random.default_rng(2)
    worst_rel = 0.0
    for _ in range(trials):
        n = int(rng.integers(2, 5))
        s = LowRankState(n, backend="chform")
        for q in range(n):
            s.clifford_1q("H", q)
        for q in range(n):
            s.t(q, dagger=bool(rng.integers(2)))
        for q in range(n - 1):
            s.cz(q, q + 1)
        exact = float(np.vdot(s.statevector(), s.statevector()).real)
        est = ne.estimate_norm2(s.terms, n, samples=4000, rng=rng)
        rel = abs(est - exact) / exact
        worst_rel = max(worst_rel, rel)
    # estimator std ~ ||psi||^2 / sqrt(L); L=4000 -> ~1.6% typical, allow margin
    assert worst_rel < 0.15, f"norm estimate off by {worst_rel:.3f}"
    print(f"[OK] Lemma 2 norm estimate -> ||psi||^2 on {trials} low-rank states "
          f"(worst rel err {worst_rel:.3f}, L=4000)")


if __name__ == "__main__":
    test_exp_sum()
    test_inner_equatorial()
    test_norm_estimate()
